### PR TITLE
fix: None LR after restoring from ckpt

### PR DESCRIPTION
**Pull request by ko3n1g**
# What does this PR do ?

After loading from checkpoint, LR looks good:

Screenshot 2026-02-08 at 16 27 51

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/training.py [hunk 1]
```diff
@@ -2949,8 +2949,16 @@ def get_e2e_base_metrics():
         for param_group in optimizer.param_groups:
             if len(param_group['params']) == 0:
                 continue
+            print(f"DEFAULT CONFIG VALUE: {param_group['default_config']}")
```
Comment (dimapihtar) [inline megatron/training/training.py]:
```suggestion
            logger.info(f"DEFAULT CONFIG VALUE: {param_group['default_config']}")
```

**Comment by dimapihtar** [comment]
let's see if @BoxiangW also approves this change.

File: megatron/training/training.py [hunk 2]
```diff
@@ -2957,10 +2957,16 @@ def get_e2e_base_metrics():
...
-            if param_group['default_config']:
+            if param_group.get('default_config', False):
                 learning_rate = param_group['lr']
+                break
+
+        if learning_rate is None:
+            for param_group in optimizer.param_groups:
+                if len(param_group['params']) > 0:
+                    learning_rate = param_group['lr']
+                    break
```
Comment (jstjohn) [inline megatron/training/training.py:2968]: This block would happen under two situations as far as I can tell. 

1. the learning rate was never updated in a param group prior to saving. In this case I suspect it would stay None. I wonder if it is safe to ask for the "lr" without a .get to support this case? For example if you call this line on a checkpoint saved before the first scheduler.step, then the LR would not be set yet unless the user overrides the LR manually through an override: https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/optimizer_param_scheduler.py#L272-L286
2. the user has overridden the min/max LR for all parameters from the default in the optimizer config. Given this we can't really know which LR is the canonical one... Maybe we could pick the "lr" that is set on most parameters in that case? Generally the canonical LR is determined by looking at whether the user overrode an lr related parameter (typically min/max LR) on a particular param group: https://github.com/NVIDIA/Megatron-LM/blob/c072f894da962453bc91a6df62a36073a1422e87/megatron/core/optimizer/__init__.py#L204 https://github.com/NVIDIA/Megatron-LM/blob/c072f894da962453bc91a6df62a36073a1422e87/megatron/core/optimizer/__init__.py#L223

File: megatron/training/training.py [hunk 3]
```diff
@@ -2957,10 +2957,16 @@ def get_e2e_base_metrics():
             params_norm = calc_params_l2_norm(model)
         learning_rate = None
         for param_group in optimizer.param_groups:
-            if len(param_group['params']) == 0:
-                continue
-            if param_group['default_config']:
+            if param_group.get('default_config', False):
```
Comment (jstjohn) [inline megatron/training/training.py:2960]: See my other comment about the case where the scheduler.step has not been called prior to the checkpoint being saved. In that case the lr would never have been set, so this could be a missing key.

File: megatron/training/training.py (hunk 2 above)
Comment (jstjohn) [reply megatron/training/training.py:2968]: This block makes me nervous. Please add a comment about why this case happens. 

Basically the goal here is to determine the canonical learning rate from the set of param groups. Typically this is done by looking at the param group with an LR schedule that has not been tweaked by the user vs the default config, and is found by looking for the `default_config=True` flag. If that is not the case then we need to pick one of the LRs and call it the canonical one. In that case the current rule in this implementation above is "pick one at random, whichever is first in the list".
